Sources/PointFree/Account/UpdateProfile.swift

```diff
@@ -4,26 +4,74 @@ import HttpPipeline
 import Prelude
 import Tuple
 
-public struct ProfileData: Codable {
+// NB: remove this `Encodable` to get a runtime crash
+public struct ProfileData: Encodable {
   public let email: EmailAddress
   public let name: String
+  public let emailSettings: [String: String]
+
+  public enum CodingKeys: String, CodingKey {
+    case email
+    case name
+    case emailSettings
+  }
+}
+
+extension ProfileData: Decodable {
+  public init(from decoder: Decoder) throws {
+    let container = try decoder.container(keyedBy: CodingKeys.self)
+
+    self.email = try container.decode(EmailAddress.self, forKey: .email)
+    self.name = try container.decode(String.self, forKey: .name)
+    self.emailSettings = (try? container.decode([String: String].self, forKey: .emailSettings)) ?? [:]
+  }
 }
 
 let updateProfileMiddleware =
-  filterMap(require2 >>> pure, or: loginAndRedirect)
-    <| { (conn: Conn<StatusLineOpen, Tuple2<ProfileData?, Database.User>>) -> IO<Conn<ResponseEnded, Data>> in
+  filterMap(require1 >>> pure, or: redirect(to: .account))
+    <<< filterMap(require2 >>> pure, or: loginAndRedirect)
+    <| { (conn: Conn<StatusLineOpen, Tuple2<ProfileData, Database.User>>) -> IO<Conn<ResponseEnded, Data>> in
       let (data, user) = lower(conn.data)
 
-      return pure(data)
-        .mapExcept(requireSome)
-        .flatMap { AppEnvironment.current.database.updateUser(user.id, $0.name, $0.email) }
+      let emailSettings = data.emailSettings.keys
+        .flatMap(Database.EmailSetting.Newsletter.init(rawValue:))
+
+      // TODO: make sure email doesn't already exist?!
+      // TODO: validate email?
+
+      let updateFlash: Middleware<HeadersOpen, HeadersOpen, Prelude.Unit, Prelude.Unit>
+      if data.email.unwrap.lowercased() != user.email.unwrap.lowercased() {
+        updateFlash = flash(.warning, "We’ve sent an email to \(data.email.unwrap) to confirm this change.")
```

Comment:
when email changes we now fire off a confirmation email and flash a message saying they have to confirm